### ISSUE TITLE
Property Couldn't be Found on Entity Exception When Accessing Key Property

### ISSUE THREAD

**Issue by malikhassan85** [dotnet/efcore#12980]
In a large query between many entities, when we split the query into subqueries and then join them using their variables, if we tried to access the Key property from a composition property in the first DBSet in the query, we get this exception: 

```
Exception: 'The property '' on entity type 'Driver' could not be found. Ensure that the property exists and has been included in the model.'

Stack trace:
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.MemberAccessBindingExpressionVisitor.GetPropertyPath(Expression expression, QueryCompilationContext queryCompilationContext, QuerySourceReferenceExpression& querySourceReferenceExpression)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.BindPropertyExpressionCore[TResult](Expression propertyExpression, IQuerySource querySource, Func`3 propertyBinder)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.BindMethodCallExpression[TResult](MethodCallExpression methodCallExpression, IQuerySource querySource, Func`3 methodCallBinder)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.BindMethodCallExpression[TResult](MethodCallExpression methodCallExpression, IQuerySource querySource, Func`4 memberBinder, Boolean bindSubQueries)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.BindMethodCallExpression[TResult](MethodCallExpression methodCallExpression, Func`4 memberBinder, Boolean bindSubQueries)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.RelationalEntityQueryableExpressionVisitor.VisitMethodCall(MethodCallExpression node)
   at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at System.Linq.Expressions.ExpressionVisitor.VisitMemberAssignment(MemberAssignment node)
   at System.Linq.Expressions.ExpressionVisitor.VisitMemberBinding(MemberBinding node)
   at System.Linq.Expressions.ExpressionVisitor.Visit[T](ReadOnlyCollection`1 nodes, Func`2 elementVisitor)
   at System.Linq.Expressions.ExpressionVisitor.VisitMemberInit(MemberInitExpression node)
   at System.Linq.Expressions.MemberInitExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.ReplaceClauseReferences(Expression expression, IQuerySource querySource, Boolean inProjection)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.VisitSelectClause(SelectClause selectClause, QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.VisitSelectClause(SelectClause selectClause, QueryModel queryModel)
   at Remotion.Linq.Clauses.SelectClause.Accept(IQueryModelVisitor visitor, QueryModel queryModel)
   at Remotion.Linq.QueryModelVisitorBase.VisitQueryModel(QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.VisitQueryModel(QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.VisitQueryModel(QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.VisitSubQueryModel(QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.LiftSubQuery(IQuerySource querySource, SubQueryExpression subQueryExpression)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.CompileJoinClauseInnerSequenceExpression(JoinClause joinClause, QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.VisitJoinClause(JoinClause joinClause, QueryModel queryModel, Int32 index)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.VisitJoinClause(JoinClause joinClause, QueryModel queryModel, Int32 index)
   at Remotion.Linq.Clauses.JoinClause.Accept(IQueryModelVisitor visitor, QueryModel queryModel, Int32 index)
   at Remotion.Linq.QueryModelVisitorBase.VisitBodyClauses(ObservableCollection`1 bodyClauses, QueryModel queryModel)
   at Remotion.Linq.QueryModelVisitorBase.VisitQueryModel(QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.VisitQueryModel(QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.VisitQueryModel(QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.CreateQueryExecutor[TResult](QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Storage.Database.CompileQuery[TResult](QueryModel queryModel)
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileQueryCore[TResult](Expression query, IQueryModelGenerator queryModelGenerator, IDatabase database, IDiagnosticsLogger`1 logger, Type contextType)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass13_0`1.<Execute>b__0()
   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQueryCore[TFunc](Object cacheKey, Func`1 compiler)
   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQuery[TResult](Object cacheKey, Func`1 compiler)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.Execute[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.Execute[TResult](Expression expression)
   at Remotion.Linq.QueryableBase`1.GetEnumerator()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at TestHost.Program.Main(String[] args) in C:\Source\Issues\EFCoreTestBed\Program.cs:line 56

```

This will happen only when something will force the query to get evaluated on memory (GetValueOrDefault() for example).

Workaround: Rather than splitting the query, write one large big query, it more be less readable, but it will work.

## Steps to produce: 

https://github.com/avinash-phaniraj-readify/EFCoreTestBed.git

Please check out branch: PropertyNotFoundException

```C#
 var driverQuery =
                    from bel in dc.Drivers
                    where bel.YearsOfExperience.GetValueOrDefault() > 1
                    select bel;

                var driverTruckInfoQuery = from truck in dc.Trucks
                               join l in driverQuery on truck.DriverId equals l.Truck.RecordId
                               select new DriveTruckInfo
                               {
                                   DepartmentId = l.Department.RecordId,
                                   DepartmentName = l.Department.Name,
                                   TruckId = truck.RecordId,
                                   TruckName = truck.Model,
                                   DriverId = l.RecordId,
                                   DriverName = l.Name
                               };

                var rgttc = from trips in dc.Trips
                            join drivertruckq in driverTruckInfoQuery on trips.DriverId equals drivertruckq.DriverId
                            select new
                            {
                                TripId = trips.RecordId,
                                TripSummary = string.Format($"Driver: {drivertruckq.DriverName}, Truck: {drivertruckq.TruckName}"),
                                trips.TotalKilometers
                            };

                var itemsList = rgttc.ToList();
```
This will reproduce the exception, however, this will work:

```C#

var driverTruckInfoQuery = from truck in dc.Trucks
                                           join l in dc.Drivers on truck.DriverId equals l.Truck.RecordId
                                           where l.YearsOfExperience.GetValueOrDefault() > 1
                                           select new DriveTruckInfo
                                           {
                                               DepartmentId = l.Department.RecordId,
                                               DepartmentName = l.Department.Name,
                                               TruckId = truck.RecordId,
                                               TruckName = truck.Model,
                                               DriverId = l.RecordId,
                                               DriverName = l.Name
                                           };

                var rgttc = from trips in dc.Trips
                            join drivertruckq in driverTruckInfoQuery on trips.DriverId equals drivertruckq.DriverId
                            select new
                            {
                                TripId = trips.RecordId,
                                TripSummary = string.Format($"Driver: {drivertruckq.DriverName}, Truck: {drivertruckq.TruckName}"),
                                trips.TotalKilometers
                            };

                var itemsList = rgttc.ToList();

```
## Further technical details:

EF Core version: 2.1.1
Database Provider: EntityFrameworkCore.SqlServerCompact35 2.1.0.1
Operating system: Win 10
IDE: Visual Studio 2017 15.7.5

**Comment by ajcvickers** [comment]
@maumar to find dupe.

**Comment by ajcvickers** [comment]
@maumar to find dupe.

**Comment by maumar** [comment]
dupe of https://github.com/aspnet/EntityFrameworkCore/issues/12575 (already fixed in 2.2)

**Comment by optiks** [comment]
Confirming that this is now fixed in 2.1.4. Many thanks team.
